File: BOJ/다이나믹 프로그래밍/11055(가장 큰 증가 부분 수열)/11055(가장 큰 증가 부분 수열).java
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.Arrays;
import java.util.StringTokenizer;

public class Main{
    static int[] sequence;
    static int[] cache;

    static int max_sum(int position, int cnt){
        int max = -1;
        if(cache[position] != -1)   return cache[position];
        if(position == cnt - 1)  return cache[position] = sequence[position];
        for(int index = position + 1; index < cnt; index++) {
            if(sequence[position] < sequence[index])    max = Math.max(max, max_sum(index, cnt));
        }
        if(max == -1)   return cache[position] = sequence[position];
        else return cache[position] = max + sequence[position];
    }

    public static void main(String[] args) throws Exception{
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int cnt = Integer.parseInt(br.readLine());
        int max_num = -1;
        sequence = new int[cnt];
        cache = new int[cnt];
        Arrays.fill(cache, -1);
        StringTokenizer st = new StringTokenizer(br.readLine(), " ");
        for(int index = 0; index < cnt; index++)    sequence[index] = Integer.parseInt(st.nextToken());
        for(int index = 0; index < cnt; index++)    max_num = Math.max(max_sum(index, cnt), max_num);
        System.out.println(max_num);
    }
}

File: BOJ/다이나믹 프로그래밍/11055(가장 큰 증가 부분 수열)/11055(가장 큰 증가 부분 수열).cpp
#include <iostream>
#include <algorithm>
#include <vector>
using namespace std;
int cnt;
vector <int> sequence;
vector <int> cache;

int max_sum(int position) {
	int max_num = -1;
	if (cache[position] != -1)	return cache[position];
	if (position == cnt - 1)	return cache[position] = sequence[position];

	for (int index = position + 1; index < cnt; index++) {
		if(sequence[position] < sequence[index])	max_num = max(max_num, max_sum(index));
	}
	if (max_num == -1)	return cache[position] = sequence[position];
	else return cache[position] = max_num + sequence[position];
}

int main() {
	ios_base::sync_with_stdio(false);	cin.tie(NULL);
	cin >> cnt;
	int max_num = -1;
	sequence.resize(cnt, 0);
	cache.resize(cnt, -1);
	for (int index = 0; index < cnt; index++)	cin >> sequence[index];
	for (int index = 0; index < cnt; index++)	max_num = max(max_sum(index), max_num);
	cout << max_num << "\n";
	return 0;
}